pkg/apis/componentconfig/types.go

```diff
@@ -160,58 +161,115 @@ const (
 	ProxyModeIPTables  ProxyMode = "iptables"
 )
 
+// SchedulerPolicyConfigMapKey defines the key of the element in the
+// scheduler's policy ConfigMap that contains scheduler's policy config.
+const SchedulerPolicyConfigMapKey string = "policy.cfg"
+
+// SchedulerPolicySourceType define ways to obtain a scheduler Policy at
+// runtime.
+type SchedulerPolicySourceType string
+
+const (
+	// SchedulePolicySourceTypeFile means get a Policy from
+	// a SchedulerPolicyFileSource.
+	SchedulePolicySourceTypeFile SchedulerPolicySourceType = "file"
+	// SchedulerPolicySourceTypeConfigMap means get a Policy from a
+	// SchedulerPolicyConfigMapSource.
+	SchedulerPolicySourceTypeConfigMap SchedulerPolicySourceType = "configmap"
+)
+
+// SchedulerPolicySource configures a means to obtain a scheduler Policy. Given
+// a SchedulerPolicySourceType, the related source should be used. Sources are
+// mutually exclusive.
+type SchedulerPolicySource struct {
+	// Type is the policy source type.
+	Type SchedulerPolicySourceType
+	// File is a file policy source.
+	File *SchedulerPolicyFileSource
+	// ConfigMap is a config map policy source.
+	ConfigMap *SchedulerPolicyConfigMapSource
+}
+
+// SchedulerPolicyFileSource is a policy serialized to disk and accessed via
+// path.
+type SchedulerPolicyFileSource struct {
+	// Path is the location of a serialized policy.
+	Path string
+}
+
+// SchedulerPolicyConfigMapSource is a policy serialized into a config map value.
+type SchedulerPolicyConfigMapSource struct {
+	// Reference is a config map containing a seralized policy in the
+	// SchedulerPolicyConfigMapKey key.
+	Reference api.ObjectReference
+}
+
+// SchedulerAlgorithmSourceType defines the supported scheduler algorithm source
+// types.
+type SchedulerAlgorithmSourceType string
+
+const (
+	// SchedulerAlgorithmSourceProvider is a named algorithm provider plugin.
+	SchedulerAlgorithmSourceProvider SchedulerAlgorithmSourceType = "provider"
+	// SchedulerAlgorithmSourcePolicy is a policy based provider.
+	SchedulerAlgorithmSourcePolicy SchedulerAlgorithmSourceType = "policy"
+)
+
+type SchedulerAlgorithmSource struct {
+	Type SchedulerAlgorithmSourceType
+	// Policy is a policy based algorithm source.
+	Policy *SchedulerPolicySource
```

Comment:
is this and provider mutually exclusive?

Reply:
Correct, the specific source fields are mutually exclusive.

Reply:
Removed Type.